particle-iot/device-os
# [Tracker] ESP32 support

### Description

This PR adds WiFi scanning support for Tracker. All other APIs except for WiFi interface state management and scanning are not available as of right now.

This feature branch should work with both the stock 0.0.6 ESP32 NCP firmware available from the factory, as well as with https://github.com/particle-iot/argon-ncp-firmware/pull/18.

### Steps to Test

- `wiring/no_fixture` tests have been fixed to allow WiFi scanning on Tracker

1. WiFi scanning should work with stock 0.0.6 ESP32 NCP firmware
2. You should be able to grab a build from https://github.com/particle-iot/argon-ncp-firmware/pull/18 and update update over serial or OTA to 0.0.7 version (`particle serial inspect` should reflect that)
3. WiFi scanning should work with 0.0.7 ESP32 NCP firmware over muxed channel
4. 0.0.7 ESP32 NCP firmware should be updatable. Manually edit version.mk in https://github.com/particle-iot/argon-ncp-firmware/pull/18 and bump version to 0.0.8 and perform a clean build. Update OTA or using serial should work.

### Example App

```c++
SYSTEM_THREAD(ENABLED);
SerialLogHandler dbg(LOG_LEVEL_ALL);

/* executes once at startup */
void setup() {
    WiFi.on();
}

/* executes continuously after setup() runs */
void loop() {
    WiFi.scan([](WiFiAccessPoint* result, void* cookie) -> void {
        char ssid[64] = {};
        memcpy(ssid, result->ssid, std::min<size_t>(sizeof(ssid) - 1, result->ssidLength));
        Log.info("Found SSID: %s (%02x:%02x:%02x:%02x:%02x:%02x)", ssid,
                result->bssid[0],
                result->bssid[1],
                result->bssid[2],
                result->bssid[3],
                result->bssid[4],
                result->bssid[5]);
    }, nullptr);
    delay(1000);
}

```

### References

- https://github.com/particle-iot/argon-ncp-firmware/pull/18
- https://github.com/particle-iot/gsm0710muxer/pull/7

---

### Completeness
